pilot/pkg/model/push_context.go

```diff
@@ -1774,6 +1776,80 @@ func (ps *PushContext) mergeGateways(proxy *Proxy) *MergedGateway {
 	return MergeGateways(out)
 }
 
+// GatewayContext contains a minimal subset of push context functionality to be exposed to GatewayAPIControllers
+type GatewayContext struct {
+	ps *PushContext
+}
+
+func NewGatewayContext(ps *PushContext) GatewayContext {
+	return GatewayContext{ps}
+}
+
+// ResolveGatewayInstances attempts to resolve all instances that a gateway will be exposed on.
+// Note: this function considers *all* instances of the service; its possible those instances will not actually be properly functioning
+// gateways, so this is not 100% accurate, but sufficient to expose intent to users.
+// The actual configuration generation is done on a per-workload basis and will get the exact set of matched instances for that workload.
+// Three sets are exposed:
+// * Internal addresses (ie istio-ingressgateway.istio-system.svc.cluster.local:80).
+// * External addresses (ie 1.2.3.4), this comes from LoadBalancer services. There may be multiple in some cases (especially multi cluster).
+// * Warnings for references that could not be resolved. These are intended to be user facing.
+func (gc GatewayContext) ResolveGatewayInstances(namespace string, gwsvcs []string, servers []*networking.Server) (internal, external, warns []string) {
+	ports := map[int]struct{}{}
+	for _, s := range servers {
+		ports[int(s.Port.Number)] = struct{}{}
+	}
+	foundInternal := sets.NewSet()
+	foundExternal := sets.NewSet()
+	warnings := []string{}
+	known := map[host.Name]struct{}{}
+	for _, g := range gwsvcs {
+		known[host.Name(g)] = struct{}{}
+		svc, f := gc.ps.ServiceIndex.HostnameAndNamespace[host.Name(g)][namespace]
+		if !f {
+			otherNamespaces := []string{}
+			for ns := range gc.ps.ServiceIndex.HostnameAndNamespace[host.Name(g)] {
+				otherNamespaces = append(otherNamespaces, `"`+ns+`"`) // Wrap in quotes for output
+			}
+			if len(otherNamespaces) > 0 {
+				sort.Strings(otherNamespaces)
+				warnings = append(warnings, fmt.Sprintf("hostname %q not found in namespace %q, but it was found in namespace(s) %v",
+					g, namespace, strings.Join(otherNamespaces, ", ")))
+			} else {
+				warnings = append(warnings, fmt.Sprintf("hostname %q not found", g))
+			}
+			continue
+		}
+		for port := range ports {
+			instances := gc.ps.ServiceIndex.instancesByPort[svc][port]
+			if len(instances) > 0 {
+				foundInternal.Insert(fmt.Sprintf("%s:%d", g, port))
+				// Fetch external IPs from all clusters
+				for _, externalIPs := range svc.Attributes.ClusterExternalAddresses {
```

Comment:
How does this work with the addresses from [Spec](https://gateway-api.sigs.k8s.io/references/spec/#networking.x-k8s.io/v1alpha1.GatewaySpec)? Do we merge them somewhere?

Reply:
The implementation is currently a user specifies the (internal) service name in the spec, and we resolve it to the (external) IP in the status. The spec is a bit vague if this is a good idea: "These addresses may differ from the addresses in the Spec, e.g. if the Gateway automatically assigns an address from a reserved pool." which implies they don't need to be the same, but doesn't specify that we can do a transformation like we do.

Using spec.Addresses to specify the Service to use is a bit of a temporary measure while we see if there is a better way to indicate it. There may not be, but I am not attached to it for sure